src/libsyntax_ext/standard_library_imports.rs

```diff
@@ -55,32 +61,88 @@ pub fn inject(
     // the prelude.
     let name = names[0];
 
-    let span = DUMMY_SP.fresh_expansion(ExpnData::allow_unstable(
-        ExpnKind::Macro(MacroKind::Attr, sym::std_inject), DUMMY_SP, edition,
-        [sym::prelude_import][..].into(),
-    ));
+    let segments = if rust_2018 {
+        [name, sym::prelude, sym::v1].iter()
+            .map(|symbol| ast::PathSegment::from_ident(ast::Ident::new(*symbol, span)))
+            .collect()
+    } else {
+        [kw::PathRoot, name, sym::prelude, sym::v1].iter()
+            .map(|symbol| ast::PathSegment::from_ident(ast::Ident::new(*symbol, call_site)))
+            .collect()
+    };
 
-    krate.module.items.insert(0, P(ast::Item {
+    let use_item = P(ast::Item {
         attrs: vec![attr::mk_attr_outer(
             attr::mk_word_item(ast::Ident::new(sym::prelude_import, span)))],
         vis: respan(span.shrink_to_lo(), ast::VisibilityKind::Inherited),
         node: ast::ItemKind::Use(P(ast::UseTree {
-            prefix: ast::Path {
-                segments: iter::once(ast::Ident::with_dummy_span(kw::PathRoot))
-                    .chain(
-                        [name, "prelude", "v1"].iter().cloned()
-                            .map(ast::Ident::from_str)
-                    ).map(ast::PathSegment::from_ident).collect(),
-                span,
-            },
+            prefix: ast::Path { segments, span },
             kind: ast::UseTreeKind::Glob,
             span,
         })),
         id: ast::DUMMY_NODE_ID,
         ident: ast::Ident::invalid(),
         span,
         tokens: None,
-    }));
+    });
+
+    let prelude_import_item = if rust_2018 {
+        let hygienic_extern_crate = P(ast::Item {
+            attrs: vec![],
+            vis: respan(span, ast::VisibilityKind::Inherited),
+            node: ast::ItemKind::ExternCrate(alt_std_name),
+            ident: ast::Ident::new(name, span),
+            id: ast::DUMMY_NODE_ID,
+            span,
+            tokens: None,
+        });
+
+        // Use an anonymous const to hide `extern crate std as hygienic_std`
```

Comment:
I don't understand why is the const necessary, can't the extern crate item already created above be used for the prelude import? It's already hygienic and not observable from outside of our "macro".

Reply:
It's not needed with the latest changes.